gr-blocks/lib/tcp_connection.cc

```diff
@@ -55,20 +55,19 @@ namespace gr {
     {
       size_t len = pmt::blob_length(vector);
 
-      // Asio async_write() requires the buffer to remain valid until the handler is called. It will be
-      // deleted in tcp_connection::handle_write.
-      char* txbuf = new char[len];
+      // Asio async_write() requires the buffer to remain valid until the handler is called.
+      boost::shared_ptr<char[]> txbuf(new char[len]);
 
       size_t temp = 0;
-      memcpy(txbuf, pmt::uniform_vector_elements(vector, temp), len);
+      memcpy(txbuf.get(), pmt::uniform_vector_elements(vector, temp), len);
 
       size_t offset = 0;
       while (offset < len) {
         // Limit the size of each write() to the MTU.
         // FIXME: Note that this has the effect of breaking a large PDU into several smaller PDUs, each
         // containing <= MTU bytes. Is this the desired behavior?
         size_t send_len = std::min((len - offset), d_buf.size());
-        boost::asio::async_write(d_socket, boost::asio::buffer(txbuf + offset, send_len),
+        boost::asio::async_write(d_socket, boost::asio::buffer(txbuf.get() + offset, send_len),
```

Comment:
I don't think this will work the way you expect; you are passing a pointer to the internal contents of a locally created shared pointer, so the actual shared pointer goes out of scope at the end of the function, and the internal raw buffer gets deleted. To take advantage of the shared pointer semantics, you need to pass the shared pointer itself to the async_write call, which will increment its reference count, then when the local 'txbuf' goes out of scope the shared pointer itself remains alive.

Reply:
You're correct we need to make sure the shared pointer stays alive until the async_write handler is called. So I make a copy of the shared pointer with `boost::bind` when creating the handler. The shared pointer's refcount will not go to zero until the write handler returns.

Reply:
Ah, got it.